### ISSUE TITLE
about left right eye landmarks question

### ISSUE THREAD

**Issue by lilinxiong** [google/mediapipe#1909]
In [all of the landmark connection vector](https://github.com/google/mediapipe/issues/1049#issuecomment-706858946) found it [FaceMesh_Map.jpg](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/mesh_map.jpg) and [Contour keypoints](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/src/keypoints.ts). The landmark for the left eye is: [left eyes](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/src/keypoints.ts#L41-L47):
![image](https://user-images.githubusercontent.com/24367027/115359053-ddc79680-a1f0-11eb-97d8-e3748197bf94.png). The landmark for the right eye is: [right eyes](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/src/keypoints.ts#L30-L36): 
![image](https://user-images.githubusercontent.com/24367027/115359307-22533200-a1f1-11eb-8c62-a71483d0032b.png)

But in [iris_tracking_gpu.pbtxt](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/iris_tracking/iris_tracking_gpu.pbtxt). define left eye boundary is (in [L84-L96](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/iris_tracking/iris_tracking_gpu.pbtxt#L84-L96)):
![image](https://user-images.githubusercontent.com/24367027/115359713-88d85000-a1f1-11eb-8156-777a4190282d.png)

define right eye boundary is (in [L98-L110](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/iris_tracking/iris_tracking_gpu.pbtxt#L98-L110)):
![image](https://user-images.githubusercontent.com/24367027/115360201-fedcb700-a1f1-11eb-917d-1fd34179a219.png)

In other words, the landmarks in ``pbtxt`` and the left and right eye landmarks defined by ``keypoints`` are inconsistent. Where is wrong?

**Comment by 04633435** [comment]
Yes, I also noticed this inconsistency when I was trying to change the range of the eye_boundary_landmarks. Did you find the correct landmarks_indices mapping yet?

**Comment by lilinxiong** [comment]
@04633435 I think [Contour keypoints](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/src/keypoints.ts) is right, because it matches the mirror image we perceive. So currently in the graph, I adjust it according to it. I haven't waited for the official reply yet, I'm still waiting for the official explanation.

**Comment by 04633435** [comment]
> @04633435 I think [Contour keypoints](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/src/keypoints.ts) is right, because it matches the mirror image we perceive. So currently in the graph, I adjust it according to it. I haven't waited for the official reply yet, I'm still waiting for the official explanation.

Thank you for your reply! 
I just look through the [Contour keypoints](https://github.com/tensorflow/tfjs-models/blob/master/facemesh/src/keypoints.ts), but there are only 229 points, half of the vertices excluding the iris landmarks. Seems like there are some points missing. Do you have any clue?

**Comment by lilinxiong** [comment]
@04633435 Face Mesh has a total of 468 points. iris is an extra 10 points(left 5, right 5). Not included in FaceMesh.
In [iris_tracking_gpu.pbtxt ConcatenateNormalizedLandmarkListCalculator](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/iris_tracking/iris_tracking_gpu.pbtxt#L158-#L163) [code](https://github.com/google/mediapipe/blob/master/mediapipe/calculators/core/concatenate_normalized_landmark_list_calculator.cc#L58-L66) can be found.
Note: left iris landmarks and right iris landmarks the order is different, you have to distinguish it yourself. I made changes in my app, so this part can’t be shared with you.

**Comment by 04633435** [comment]
Hi, @lilinxiong, I solved my problem. And for the problem you mentioned above, in [L84-L96](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/iris_tracking/iris_tracking_gpu.pbtxt#L84-L96), it selects out the inner corner and outer corner landmarks of the left eye, and within [the subgraph iris_landmark_landmarks_to_roi.pbtxt](https://github.com/google/mediapipe/blob/master/mediapipe/modules/iris_landmark/iris_landmark_landmarks_to_roi.pbtxt) of [the iris_landmark module](https://github.com/google/mediapipe/tree/master/mediapipe/modules/iris_landmark), these two landmarks are used to create a square boundary around the left eye area. So in [L84-L96](https://github.com/google/mediapipe/blob/master/mediapipe/graphs/iris_tracking/iris_tracking_gpu.pbtxt#L84-L96), there are only landmark 33 and 133, not the whole left eye contours landmarks.

**Comment by lilinxiong** [comment]
@04633435 I don’t think we are talking about the same issue.
```
Face Mesh has a total of 468 points. iris is an extra 10 points(left 5, right 5). ........
```
My answer here is that the iris coordinates are added to the content related to FaceMesh landmark.
Including the order of iris coordinates and the position added to FaceMesh.
```
And for the problem you mentioned above, in L84-L96, it selects out the inner corner and outer corner landmarks of the left eye.....
```
As for what you are answering now, I think you only need to exchange the subscript values of left and right in the original pbtxt.
The original intent of this issue is that the left and right in pbtxt are written inversely, so even after swapping left and right, the result is still the opposite. You can verify it yourself.
I'm not sure, whether you know what I want to express, but the final conclusion is: even after you swap left and right, their final value is unchanged, just changing here will not take effect. Other places need to be changed.

**Comment by sureshdagooglecom** [comment]
Hi @lilinxiong ,
Are you still looking for the resolution on this issue.

**Comment by kuaashish** [comment]
Hello @lilinxiong,
We are upgrading the [MediaPipe Legacy Solutions](https://developers.google.com/mediapipe/solutions/guide#legacy) to [new MediaPipe solutions](https://developers.google.com/mediapipe/solutions) However, the libraries, documentation, and source code for all the MediaPipe Legacy Solutions will continue to be available in our [GitHub repository](https://github.com/google/mediapipe) and through library distribution services, such as [Maven](https://maven.google.com/web/index.html?#com.google.mediapipe) and [NPM](https://www.npmjs.com/search?q=%40mediapipe).

You can continue to use those legacy solutions in your applications if you choose. Though, we would request you to check [new MediaPipe solutions](https://developers.google.com/mediapipe/solutions) which can help you more easily build and customize ML solutions for your applications. These new solutions will provide a superset of capabilities available in the legacy solutions.
